controller/api/destination/watcher/endpoints_watcher.go

```diff
@@ -283,6 +298,81 @@ func (ew *EndpointsWatcher) deleteEndpoints(obj interface{}) {
 	}
 }
 
+func (ew *EndpointsWatcher) addEndpointSlice(obj interface{}) {
+	es, ok := obj.(*discovery.EndpointSlice)
+	if !ok {
+		ew.log.Errorf("error processing EndpointSlice resource, got %T expected *discovery.EndpointSlice", obj)
+		return
+	}
+
+	if es.Namespace == kubeSystem {
+		return
+	}
+	id, err := getEndpointSliceServiceID(es)
+	if err != nil {
+		ew.log.Errorf("Could not fetch service name for EndpointSlice [%s/%s]:%v", es.Namespace, es.Name, err)
+		return
+	}
+	sp := ew.getOrNewServicePublisher(id)
+	sp.addEndpointSlice(es)
+}
+
+func (ew *EndpointsWatcher) updateEndpointSlice(oldObj interface{}, newObj interface{}) {
+	oldSlice, ok := oldObj.(*discovery.EndpointSlice)
+	if !ok {
+		ew.log.Errorf("error processing EndpointSlice resource, got %#v expected *discovery.EndpointSlice", oldObj)
+		return
+	}
+	newSlice, ok := newObj.(*discovery.EndpointSlice)
+	if !ok {
+		ew.log.Errorf("error processing EndpointSlice resource, got %#v expected *discovery.EndpointSlice", oldObj)
+		return
+	}
```

Comment:
I think you can refactor that logic in a separate method since you are repeating it here and in add as well. Also you are formatting with `%#v` while using `%T` in the add method

Reply:
Hmm, I looked at this based on your suggestion, but I'm not sure how well it can be extracted out into a separate function without losing context or having the same amount of lines in. I've left some thoughts below :) 

* If we extract out just the type casting, we would not save much in terms of space since the new function would still have to return an error or a bool in case the obj is not a slice; if we would not do that we can't exit early from the handler method.
* Although add and update work very similarily, for `add` we use `getOrNewServicePublisher`, for `update` we use `getServicePublisher`. This subtle difference makes it harder to extract the code.
* A third alternative is to create a function which returns a `*discovery.EndpointSlice` resource and a `ServiceID`, this is the approach I took. If the function does not fit well, I can take it out, however.

Let me know what you think.

Reply:
If that is the case, it makes sense to leave it as  is

Reply:
Cool, will roll back the changes in this case